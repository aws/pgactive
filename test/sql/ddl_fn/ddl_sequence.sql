--  ALTER TABLE public.DROP COLUMN (pk column)
SELECT bdr.bdr_replicate_ddl_command($DDL$ CREATE TABLE public.test (test_id SERIAL); $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);

\c postgres
\d+ test
SELECT relname, relkind FROM pg_class WHERE relname = 'test_test_id_seq';
\d+ test_test_id_seq

SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER TABLE public.test  DROP COLUMN test_id; $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\c regression
\d+ test
SELECT relname, relkind FROM pg_class WHERE relname = 'test_test_id_seq';

SELECT bdr.bdr_replicate_ddl_command($DDL$ DROP TABLE public.test; $DDL$);

-- ADD CONSTRAINT PRIMARY KEY
SELECT bdr.bdr_replicate_ddl_command($DDL$ CREATE TABLE public.test (test_id SERIAL NOT NULL); $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\c postgres
\d+ test
SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER TABLE public.test ADD CONSTRAINT test_pkey PRIMARY KEY (test_id); $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\c regression
\d+ test

SELECT bdr.bdr_replicate_ddl_command($DDL$ DROP TABLE public.test; $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\c postgres

-- normal sequence
SELECT bdr.bdr_replicate_ddl_command($DDL$ CREATE SEQUENCE public.test_seq increment 10; $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\d+ test_seq
\c postgres
\d+ test_seq

SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq increment by 10; $DDL$);
SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq minvalue 0; $DDL$);
SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq maxvalue 1000000; $DDL$);
SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq restart; $DDL$);
SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq cache 10; $DDL$);
SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq cycle; $DDL$);
SELECT bdr.bdr_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq RENAME TO renamed_test_seq; $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\d+ test_seq
\d+ renamed_test_seq
\c regression
\d+ test_seq
\d+ renamed_test_seq
\c postgres


SELECT bdr.bdr_replicate_ddl_command($DDL$ DROP SEQUENCE public.renamed_test_seq; $DDL$);

SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\d+ renamed_test_seq;
\c regression
\d+ renamed_test_seq

SELECT bdr.bdr_replicate_ddl_command($DDL$ CREATE SEQUENCE public.test_seq; $DDL$);
-- DESTINATION COLUMN TYPE REQUIRED BIGINT
SELECT bdr.bdr_replicate_ddl_command($DDL$ DROP TABLE IF EXISTS public.test_tbl; $DDL$);
SELECT bdr.bdr_replicate_ddl_command($DDL$ CREATE TABLE public.test_tbl (a int DEFAULT bdr.bdr_snowflake_id_nextval('public.test_seq'),b text); $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\d+ test_tbl
\c postgres
\d+ test_tbl
INSERT INTO test_tbl(b) VALUES('abc');
SELECT count(*) FROM test_tbl;

SELECT bdr.bdr_replicate_ddl_command($DDL$ DROP TABLE public.test_tbl; $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);

SELECT bdr.bdr_replicate_ddl_command($DDL$ CREATE TABLE public.test_tbl (a bigint DEFAULT bdr.bdr_snowflake_id_nextval('public.test_seq'),b text); $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\d+ test_tbl
\c postgres
\d+ test_tbl
INSERT INTO test_tbl(b) VALUES('abc');
SELECT count(*) FROM test_tbl;
SELECT bdr.bdr_replicate_ddl_command($DDL$ DROP SEQUENCE public.test_seq CASCADE; $DDL$);
SELECT bdr.bdr_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\d+ test_tbl
\c regression
\d+ test_tbl
