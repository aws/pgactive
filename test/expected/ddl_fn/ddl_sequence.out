--  ALTER TABLE public.DROP COLUMN (pk column)
SELECT pgactive.pgactive_replicate_ddl_command($DDL$ CREATE TABLE public.test (test_id SERIAL); $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\c postgres
\d+ test
                                                   Table "public.test"
 Column  |  Type   | Collation | Nullable |                Default                | Storage | Stats target | Description 
---------+---------+-----------+----------+---------------------------------------+---------+--------------+-------------
 test_id | integer |           | not null | nextval('test_test_id_seq'::regclass) | plain   |              | 

SELECT relname, relkind FROM pg_class WHERE relname = 'test_test_id_seq';
     relname      | relkind 
------------------+---------
 test_test_id_seq | S
(1 row)

\d+ test_test_id_seq
                  Sequence "public.test_test_id_seq"
  Type   | Start | Minimum |  Maximum   | Increment | Cycles? | Cache 
---------+-------+---------+------------+-----------+---------+-------
 integer |     1 |       1 | 2147483647 |         1 | no      |     1
Owned by: public.test.test_id

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER TABLE public.test  DROP COLUMN test_id; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\c regression
\d+ test
                                  Table "public.test"
 Column | Type | Collation | Nullable | Default | Storage | Stats target | Description 
--------+------+-----------+----------+---------+---------+--------------+-------------

SELECT relname, relkind FROM pg_class WHERE relname = 'test_test_id_seq';
 relname | relkind 
---------+---------
(0 rows)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ DROP TABLE public.test; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

-- ADD CONSTRAINT PRIMARY KEY
SELECT pgactive.pgactive_replicate_ddl_command($DDL$ CREATE TABLE public.test (test_id SERIAL NOT NULL); $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\c postgres
\d+ test
                                                   Table "public.test"
 Column  |  Type   | Collation | Nullable |                Default                | Storage | Stats target | Description 
---------+---------+-----------+----------+---------------------------------------+---------+--------------+-------------
 test_id | integer |           | not null | nextval('test_test_id_seq'::regclass) | plain   |              | 

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER TABLE public.test ADD CONSTRAINT test_pkey PRIMARY KEY (test_id); $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\c regression
\d+ test
                                                   Table "public.test"
 Column  |  Type   | Collation | Nullable |                Default                | Storage | Stats target | Description 
---------+---------+-----------+----------+---------------------------------------+---------+--------------+-------------
 test_id | integer |           | not null | nextval('test_test_id_seq'::regclass) | plain   |              | 
Indexes:
    "test_pkey" PRIMARY KEY, btree (test_id)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ DROP TABLE public.test; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\c postgres
-- normal sequence
SELECT pgactive.pgactive_replicate_ddl_command($DDL$ CREATE SEQUENCE public.test_seq increment 10; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\d+ test_seq
                          Sequence "public.test_seq"
  Type  | Start | Minimum |       Maximum       | Increment | Cycles? | Cache 
--------+-------+---------+---------------------+-----------+---------+-------
 bigint |     1 |       1 | 9223372036854775807 |        10 | no      |     1

\c postgres
\d+ test_seq
                          Sequence "public.test_seq"
  Type  | Start | Minimum |       Maximum       | Increment | Cycles? | Cache 
--------+-------+---------+---------------------+-----------+---------+-------
 bigint |     1 |       1 | 9223372036854775807 |        10 | no      |     1

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq increment by 10; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq minvalue 0; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq maxvalue 1000000; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq restart; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq cache 10; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq cycle; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ ALTER SEQUENCE public.test_seq RENAME TO renamed_test_seq; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\d+ test_seq
\d+ renamed_test_seq
                Sequence "public.renamed_test_seq"
  Type  | Start | Minimum | Maximum | Increment | Cycles? | Cache 
--------+-------+---------+---------+-----------+---------+-------
 bigint |     1 |       0 | 1000000 |        10 | yes     |    10

\c regression
\d+ test_seq
\d+ renamed_test_seq
                Sequence "public.renamed_test_seq"
  Type  | Start | Minimum | Maximum | Increment | Cycles? | Cache 
--------+-------+---------+---------+-----------+---------+-------
 bigint |     1 |       0 | 1000000 |        10 | yes     |    10

\c postgres
SELECT pgactive.pgactive_replicate_ddl_command($DDL$ DROP SEQUENCE public.renamed_test_seq; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\d+ renamed_test_seq;
\c regression
\d+ renamed_test_seq
SELECT pgactive.pgactive_replicate_ddl_command($DDL$ CREATE SEQUENCE public.test_seq; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

-- DESTINATION COLUMN TYPE REQUIRED BIGINT
SELECT pgactive.pgactive_replicate_ddl_command($DDL$ DROP TABLE IF EXISTS public.test_tbl; $DDL$);
NOTICE:  table "test_tbl" does not exist, skipping
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ CREATE TABLE public.test_tbl (a int DEFAULT pgactive.pgactive_snowflake_id_nextval('public.test_seq'),b text); $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\d+ test_tbl
                                                       Table "public.test_tbl"
 Column |  Type   | Collation | Nullable |                      Default                       | Storage  | Stats target | Description 
--------+---------+-----------+----------+----------------------------------------------------+----------+--------------+-------------
 a      | integer |           |          | pgactive.pgactive_snowflake_id_nextval('test_seq'::regclass) | plain    |              | 
 b      | text    |           |          |                                                    | extended |              | 

\c postgres
\d+ test_tbl
                                                       Table "public.test_tbl"
 Column |  Type   | Collation | Nullable |                      Default                       | Storage  | Stats target | Description 
--------+---------+-----------+----------+----------------------------------------------------+----------+--------------+-------------
 a      | integer |           |          | pgactive.pgactive_snowflake_id_nextval('test_seq'::regclass) | plain    |              | 
 b      | text    |           |          |                                                    | extended |              | 

INSERT INTO test_tbl(b) VALUES('abc');
ERROR:  integer out of range
SELECT count(*) FROM test_tbl;
 count 
-------
     0
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ DROP TABLE public.test_tbl; $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ CREATE TABLE public.test_tbl (a bigint DEFAULT pgactive.pgactive_snowflake_id_nextval('public.test_seq'),b text); $DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\d+ test_tbl
                                                       Table "public.test_tbl"
 Column |  Type  | Collation | Nullable |                      Default                       | Storage  | Stats target | Description 
--------+--------+-----------+----------+----------------------------------------------------+----------+--------------+-------------
 a      | bigint |           |          | pgactive.pgactive_snowflake_id_nextval('test_seq'::regclass) | plain    |              | 
 b      | text   |           |          |                                                    | extended |              | 

\c postgres
\d+ test_tbl
                                                       Table "public.test_tbl"
 Column |  Type  | Collation | Nullable |                      Default                       | Storage  | Stats target | Description 
--------+--------+-----------+----------+----------------------------------------------------+----------+--------------+-------------
 a      | bigint |           |          | pgactive.pgactive_snowflake_id_nextval('test_seq'::regclass) | plain    |              | 
 b      | text   |           |          |                                                    | extended |              | 

INSERT INTO test_tbl(b) VALUES('abc');
SELECT count(*) FROM test_tbl;
 count 
-------
     1
(1 row)

SELECT pgactive.pgactive_replicate_ddl_command($DDL$ DROP SEQUENCE public.test_seq CASCADE; $DDL$);
NOTICE:  drop cascades to default value for column a of table public.test_tbl
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
----------------------------------------
 
(1 row)

\d+ test_tbl
                                 Table "public.test_tbl"
 Column |  Type  | Collation | Nullable | Default | Storage  | Stats target | Description 
--------+--------+-----------+----------+---------+----------+--------------+-------------
 a      | bigint |           |          |         | plain    |              | 
 b      | text   |           |          |         | extended |              | 

\c regression
\d+ test_tbl
                                 Table "public.test_tbl"
 Column |  Type  | Collation | Nullable | Default | Storage  | Stats target | Description 
--------+--------+-----------+----------+---------+----------+--------------+-------------
 a      | bigint |           |          |         | plain    |              | 
 b      | text   |           |          |         | extended |              | 

